### PR TITLE
spk ring create

### LINKED ISSUE #969: Add a ring in SPK
**As a:**
spk user
**I want:**
to be able to add a ring
**So that:**
i can have rings
**Describe the solution you'd like:**
a new `spk ring create <ring-name>` command to create a ring in a spk managed project.

**Acceptance Criteria:**
- [x] Create a new command `spk ring create <ring-name>`
- [x] Command adds the ring to the `bedrock.yaml`
- [x] For all services linked in `bedrock.yaml`, add the new ring to the trigger branches in the service's build pipeline

**Does this require updates to documentation?:**
Yes

### PR DESCRIPTION
**Pull request by mtarng**
Closes https://github.com/microsoft/bedrock/issues/969

### CODE REVIEW

File: src/commands/ring/create.ts [hunk 1]
```diff
@@ -20,18 +24,31 @@ export const execute = async (
...
+    logger.info(`Project path: ${projectPath}`);
 
-    checkDependencies(projectPath);
+    checkDependencies(projectPath, ringName);
 
-    // Check if ring already exists, if it does, warn and exit
     // Add ring to bedrock.yaml
+    addNewRing(projectPath, ringName);
     // Add ring to all linked service build pipelines' branch triggers
+    const bedrockFile: IBedrockFile = loadBedrockFile(projectPath);
```
Comment (edaena) [inline src/commands/ring/create.ts]: use `Bedrock(projectPath);` from config.ts

**Review by edaena — APPROVED** [review]

File: src/commands/ring/create.test.ts [hunk 2]
```diff
@@ -12,19 +16,61 @@ afterAll(() => {
...
-  it("negative test", async () => {
+describe("checkDependencies", () => {
+  it("Project not initialized, it should fail.", async () => {
+    try {
+      const tmpDir = createBedrockYaml();
+      checkDependencies(tmpDir, "");
+      expect(true).toBe(false); // Should not reach here
+    } catch (e) {
+      expect(e).not.toBeNull();
+    }
```
Comment (dennisseah) [inline src/commands/ring/create.test.ts]: can be
```
const tmpDir = createBedrockYaml();
expect(checkDependencies(tmpDir, "")).toThrow();
```
```

File: src/commands/ring/create.test.ts [hunk 3]
```diff
@@ -12,19 +16,61 @@ afterAll(() => {
...
+          }
+        },
+        services: {},
+        variableGroups: ["testvg"]
+      });
+      checkDependencies(tmpDir, "not-master");
+      expect(true).toBe(true);
+    } catch (e) {
+      expect(true).toBe(false); // Should not reach here
+    }
```
Comment (dennisseah) [inline src/commands/ring/create.test.ts]: can be just
```
      const tmpDir = createBedrockYaml();
      createBedrockYaml(tmpDir, {
        rings: {
          master: {
            isDefault: true
          }
        },
        services: {},
        variableGroups: ["testvg"]
      });
      checkDependencies(tmpDir, "not-master");
```
test will fail if exception is thrown

File: src/lib/fileutils.test.ts [hunk 4]
```diff
@@ -239,6 +240,79 @@ describe("generateServiceBuildAndUpdatePipelineYaml", () => {
...
+      updateTriggerBranchesForServiceBuildAndUpdatePipeline(
+        newRings,
+        path.join(targetDirectory, serviceDirectory)
+      );
+    } catch (e) {
+      expect(true).toBe(false); // Should not reach here
+    }
+
+    const absTargetPath = path.resolve(targetDirectory);
+    const expectedFilePath = `${absTargetPath}/${serviceDirectory}/${SERVICE_PIPELINE_FILENAME}`;
```
Comment (dennisseah) [inline src/lib/fileutils.test.ts]: i am not sure. I think this should be
```
const expectedFilePath = path.join(absTargetPath, serviceDirectory, SERVICE_PIPELINE_FILENAME);
 
```

File: src/commands/ring/create.ts [hunk 5]
```diff
@@ -54,9 +71,17 @@ export const commandDecorator = (command: commander.Command): void => {
...
   const fileInfo: IBedrockFileInfo = bedrockFileInfo(projectPath);
   if (fileInfo.exist === false) {
     throw new Error(PROJECT_INIT_DEPENDENCY_ERROR_MESSAGE);
   }
+
+  // Check if ring already exists, if it does, warn and exit
+  const bedrockFile: IBedrockFile = loadBedrockFile(projectPath);
+  if (ringName in bedrockFile.rings) {
+    throw new Error(
+      `ring: ${ringName} already exists in project bedrock.yaml.`
```
Comment (dennisseah) [inline src/commands/ring/create.ts]: can replace `bedrock.yaml` with the constant value defined in constants.ts

File: src/commands/ring/index.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import { Command } from "../command";
 
-const subcommands = ["add", "delete", "set-default"];
+const subcommands = ["create", "delete", "set-default"];
```
Comment (andrebriggs) [inline src/commands/ring/index.ts:3]: @mtarng does the documentation reflect this name change?

Comment (evanlouie) [reply src/commands/ring/index.ts:3]: to my understanding, all docs (at least the ones @bnookala and I wrote) referenced `create` to begin with -- and the `create.md` for rings mentions `create`

**Review by evanlouie — APPROVED** [review]

File: src/commands/ring/index.ts (hunk 6 above)
Comment (mtarng) [reply src/commands/ring/index.ts:3]: Yes. All docs refer to `spk ring create`. I had originally created the command as `spk ring add`, but this corrects it.

**Pull request merged**